File: easy/symmetric-tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isSymmetric(root):
    """
    Keep track of each tree side, and compare.
    """
    # Single node, should be true.
    if not root.left and not root.right:
        return True
    if (root.left and not root.right) or (root.right and not root.left):
        return False

    left_res = []
    left_curr = root.left
    left_stack = []

    while left_curr is not None or len(left_stack) > 0:
        while left_curr is not None:
            left_stack.append(left_curr)
            left_curr = left_curr.left
        left_curr = left_stack.pop()
        # In res, also have to keep track of depth, so each entry is tuple of
        # value and the current depth in the stack.
        left_res.append((left_curr.val, len(left_stack)))
        left_curr = left_curr.right

    right_res = []
    right_curr = root.right
    right_stack = []

    while right_curr is not None or len(right_stack) > 0:
        while right_curr is not None:
            right_stack.append(right_curr)
            right_curr = right_curr.right
        right_curr = right_stack.pop()
        right_res.append((right_curr.val, len(right_stack)))
        right_curr = right_curr.left

    if right_res == left_res:
        return True
    else:
        return False


import unittest

class TestIsSymmetric(unittest.TestCase):

    def test_first_sample(self):
        root = TreeNode(
                val=1,
                left=TreeNode(
                    val=2,
                    left=TreeNode(
                        val=3,
                        left=None,
                        right=None),
                    right=TreeNode(
                        val=4,
                        left=None,
                        right=None)
                    ),
                right=TreeNode(
                    val=2,
                    left=TreeNode(
                        val=4,
                        left=None,
                        right=None),
                    right=TreeNode(
                        val=3,
                        left=None,
                        right=None)
                    )
                )
        self.assertEqual(isSymmetric(root), True)

    def test_second_sample(self):
        root = TreeNode(
                val=1,
                left=TreeNode(
                    val=2,
                    left=None,
                    right=TreeNode(
                        val=3,
                        left=None,
                        right=None)
                    ),
                right=TreeNode(
                    val=2,
                    left=None,
                    right=TreeNode(
                        val=3,
                        left=None,
                        right=None)
                    )
                )
        self.assertEqual(isSymmetric(root), False)

    def test_third_sample(self):
        root = TreeNode(val=1, left=None, right=None)
        self.assertEqual(isSymmetric(root), True)

    def test_fourth_sample(self):
        root = TreeNode(
                val=1,
                left=TreeNode(
                    val=2,
                    left=TreeNode(
                        val=2,
                        left=None,
                        right=None),
                    right=None),
                right=TreeNode(
                    val=2,
                    left=TreeNode(
                        val=2,
                        left=None,
                        right=None),
                    right=None)
                )
        self.assertEqual(isSymmetric(root), False)

    def test_fifth_sample(self):
        root = TreeNode(
                val=5,
                left=TreeNode(
                    val=4,
                    left=None, 
                    right=TreeNode(
                        val=1, 
                        left=TreeNode(
                            val=2,
                            left=None,
                            right=None
                            ),
                        right=None
                        )
                    ),
                right=TreeNode(
                    val=1,
                    left=None,
                    right=TreeNode(
                        val=4,
                        left=TreeNode(
                            val=2,
                            left=None,
                            right=None
                            ),
                        right=None
                        )
                    )
                )
        self.assertEqual(isSymmetric(root), False)


if __name__ == '__main__':
    unittest.main()

File: easy/symmetric-tree.cpp
#include <iostream>
#include <stack>
#include <vector>
#include <cassert>

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int val = 0, TreeNode* left = nullptr, TreeNode* right = nullptr)
        : val(val), left(left), right(right) {}
};

bool isSymmetric(TreeNode* root) {
    if (!root->left && !root->right) {
        return true;
    }
    if ((root->left && !root->right) || (root->right && !root->left)) {
        return false;
    }

    std::vector<std::pair<int, int>> left_res;
    TreeNode* left_curr = root->left;
    std::stack<TreeNode*> left_stack;

    while (left_curr != nullptr || !left_stack.empty()) {
        while (left_curr != nullptr) {
            left_stack.push(left_curr);
            left_curr = left_curr->left;
        }
        left_curr = left_stack.top();
        left_stack.pop();
        left_res.push_back({left_curr->val, left_stack.size()});
        left_curr = left_curr->right;
    }

    std::vector<std::pair<int, int>> right_res;
    TreeNode* right_curr = root->right;
    std::stack<TreeNode*> right_stack;

    while (right_curr != nullptr || !right_stack.empty()) {
        while (right_curr != nullptr) {
            right_stack.push(right_curr);
            right_curr = right_curr->right;
        }
        right_curr = right_stack.top();
        right_stack.pop();
        right_res.push_back({right_curr->val, right_stack.size()});
        right_curr = right_curr->left;
    }

    return right_res == left_res;
}

int main() {
    // Test 1
    {
        TreeNode* root = new TreeNode(
            1,
            new TreeNode(2, new TreeNode(3), new TreeNode(4)),
            new TreeNode(2, new TreeNode(4), new TreeNode(3))
        );
        assert(isSymmetric(root) == true);
        delete root;
    }

    // Test 2
    {
        TreeNode* root = new TreeNode(1);
        assert(isSymmetric(root) == true);
        delete root;
    }

    // Test 3
    {
        TreeNode* root = new TreeNode(
            1,
            new TreeNode(2, new TreeNode(2)),
            new TreeNode(2, new TreeNode(2))
        );
        assert(isSymmetric(root) == false);
        delete root;
    }

    std::cout << "All tests passed successfully.\n";
    return 0;
}

